File: KISIM2/src/KimlikNo.c
/**
*
* @author Burak Yazan - burak.yazan@ogr.sakarya.edu.tr
* @since  Nisan - 2020
* <p>
* KimlikNo.c KimlikNo baslik dosyasinin gerceklendigi kaynak dosyasidir.
* </p>
*/

#include "KimlikNo.h"

// KimlikNo yapisindan hafizada alan olusturup bunu donduren fonksiyon
KimlikNo kimlikNoOlustur() {
	KimlikNo this = (KimlikNo)malloc(sizeof(struct KIMLIKNO)); // KimlikNo yapisi icin hafizada yer ayriliyor.
	
	this->numaraUret = &_kimlikNoUret; // numaraUret fonksiyon gostericisine fonksiyonun adresi ataniyor.
	this->yoket = &_kimlikNoYoket; // yoket fonksiyon gostericisine fonksiyonun adresi ataniyor.

	this->numaraUret(this); // kimlik numarasi olusturulmak icin fonksiyon cagriliyor
	return this; // yapi adresi donduruluyor
}

// kimlik no kurallarına gore kimlik numarasi olusturan fonksiyon
void _kimlikNoUret(KimlikNo _kimlikNo) {
	Rastgele rastgele = rastgeleOlustur(10); // rastgele sayi uretmek icin olusturuyoruz

	char** ptr = &_kimlikNo->numara;
	*ptr = (char*)malloc(sizeof(char) * 12); // kimlik numarasi alani icin yer ayriliyor

	char kimlikNo[12] = { '\0' }; // kimlik numarasi icin bir dizi olusturuyoruz

	// kimlik numaralarinin ilk hanesi 0 olamaz
	// bu yuzden ilk hane 0 olmayana kadar rastgele rakam atiyoruz
	sprintf(kimlikNo, "%d", rastgele->sayiUret(rastgele));
	while (kimlikNo[0] == '0') {
		sprintf(kimlikNo, "%d", rastgele->sayiUret(rastgele)); 
	}

	// ilk haneden sonra geri kalan 8 hane icin rastgele rakamlar atiyoruz
	for (int i = 1; i < 9; i++) {
		sprintf(kimlikNo + i, "%d", rastgele->sayiUret(rastgele)); 
	}

	int cift_basamak = 0; // cift siradaki rakamlarin toplamini tutmak icin degisken
	int tek_basamak = 0; // tek siradaki rakamlarin toplamini tutmak icin degisken 
	int tum_basamaklar = 0; // tum rakamlarin toplamini tutmak icin degisken 
	for (int i = 0; i < 9; i++) { // ilk 9 hane icin dizi uzerinde donuyoruz
		if (((i + 1) % 2) == 0) { // cift siradaki sayilar cift_basamak ta toplaniyor
			cift_basamak += (int)kimlikNo[i] - 48;
		}
		else { // tek siradaki sayilar tek_basamak ta toplaniyor
			tek_basamak += (int)kimlikNo[i] - 48;
		} // tum sayilar tum_basamak ta toplaniyor
		tum_basamaklar += (int)kimlikNo[i] - 48;
	}

	// kimlik numarasinin 10. hanesini matematiksel olarak hesapliyoruz ve diziye atiyoruz
	sprintf(kimlikNo + 9, "%d", ((((tek_basamak) * 7) - (cift_basamak)) % 10));
	// hesapladigimiz 10.haneyi de tum_basamaklar toplamina ekliyoruz
	tum_basamaklar += (int)kimlikNo[9] - 48;
	// kimlik numarasinin 11. hanesi de matematiksel olarak hesaplaniyor
	sprintf(kimlikNo + 10, "%d", (tum_basamaklar % 10));

	strcpy(*ptr, kimlikNo); // dizimizdeki elemanlari pointerin gosterdigi alana kopyaliyoruz
	rastgele->yoket(rastgele); // hafizadan aldigimiz alani geri birakiyoruz
}

// uretilen kimlik numalarini kontrol etmemizi saglayan fonksiyon
int kimlikKontrol(char* kimlikNo) {
	// eger kimlik numarasinin ilk hanesi 0 ise kontrol etmeye gerek yok kimlik numarasi yanlistir
	if (kimlikNo[0] == '0') {
		return 0;
	}

	int cift_basamak = 0; // cift siradaki rakamlarin toplamini tutmak icin degisken
	int tek_basamak = 0; // tek siradaki rakamlarin toplamini tutmak icin degisken 
	int tum_basamaklar = 0; // tum rakamlarin toplamini tutmak icin degisken 
	for (int i = 0; i < 9; i++) { // ilk 9 hane icin dizi uzerinde donuyoruz
		if (((i + 1) % 2) == 0) { // cift siradaki sayilar cift_basamak ta toplaniyor
			cift_basamak += (int)kimlikNo[i] - 48;
		}
		else { // tek siradaki sayilar tek_basamak ta toplaniyor
			tek_basamak += (int)kimlikNo[i] - 48;
		} // tum sayilar tum_basamak ta toplaniyor
		tum_basamaklar += (int)kimlikNo[i] - 48;
	}

	// kimlik numarasinin 10. hanesini matematiksel olarak hesapliyoruz ve diziye atiyoruz
	int onBasamak = ((((tek_basamak) * 7) - (cift_basamak)) % 10);
	// hesaplanan 10. haneyi de tum_basamaklar toplamina ekliyoruz
	tum_basamaklar += onBasamak;
	// kimlik numarasinin son hanesini hesapliyoruz
	int sonBasamak = tum_basamaklar % 10;

	// eger bizim hesapladigimiz 10. basamak ile kimlik numarasinin 10.basamagi ayni degilse kimlik no yanlistir
	if (kimlikNo[9] != onBasamak + 48) {
		return 0;
	} // eger bizim hesapladigimiz son basamak ile kimlik numarasinin 10. basamagi ayni degilse kimlik no yanlistir
	else if (kimlikNo[10] != sonBasamak + 48) {
		return 0;
	}

	// eger program buraya kadar geldiyse kimlik numarasi dogrudur bu yuzden 1 donduruyoruz
	return 1;
}

// hafizadan alinan alani geri birakmaya yarayan fonksiyon
void _kimlikNoYoket(KimlikNo this) {
	if (this != NULL) {  // eger kimlik no alani bos degilse bunu serbest birakicaz
		free(this->numara); // once kimlik alanina ait olan numara alanini serbest birakiyoruz.
		free(this); // kimlik no alani serbest birakiliyor.
		this = NULL;
	}
}

File: KISIM1/Odev2Kutuphane/src/RASTGELEKISIURET/KimlikNo.java
/**
*
* @author Burak Yazan - burak.yazan@ogr.sakarya.edu.tr
* @since  Nisan - 2020
* <p>
* KimlikNo sinifi kimlik numaralari olusturmaniza ve varolan kimlik numalarini kontrol etmenizi saglar.
* </p>
*/
package RASTGELEKISIURET;

public class KimlikNo {
    private String kimlikNo; // kimlik numarasini tutmak icin ozel bir string degiskeni tanimladim.
    
    /*
     * KimlikNo sinifi icin kurucu fonksiyon
     */
    public KimlikNo(){
        kimlikNo = "";
        setKimlikNo(); // nesne olusturuldugunda kimlik numarasi olusturuluyor.
    }
    
    /*
     * Kurallara uygun olarak kimlik numarasi uretmek icin kullanilan fonksiyondur.
     */
    public void setKimlikNo(){
        int[] kimlikNoArr = new int[11]; // kimlik numalari 11 haneli oldugu icin 11 haneli bir integer dizisi olusturuyorum.
        
        for(int i=0;i<9;i++){ // kimlik numarasinin ilk 9 hanesi icin rastgele rakamlar atiyorum.
            kimlikNoArr[i] = Rastgele.sonraki(10);
        }
        
        // kimlik numalarinin ilk numarasi 0 olamaz
        while(kimlikNoArr[0] == 0){ // kimlik numarasi 0 olmayana kadar tekrar atama yapiyorum.
            kimlikNoArr[0] = Rastgele.sonraki(10);
        }
        
        int cift_basamak = 0; // cift basamaktaki rakamlarin toplamini tutmak icin olusturulan degisken. 
        int tek_basamak = 0; // tek basamaktaki rakamlarin toplamini tutmak icin olusturulan degisken.
        int tum_basamaklar = 0; // tum basamaklardaki rakamlarin toplamini tutmak icin olusturulan degisken
        for(int i=0;i<9;i++){ // ilk 9 hanesi icin islem yapiyoruz.
            if(((i+1)%2) == 0){ // cift hanelerdeki rakamlari cift_basamak degiskeninde topluyoruz.
                cift_basamak += kimlikNoArr[i]; 
            } else { // tek hanelerdeki rakamlari tek_basamak degiskeninde topluyoruz.
                tek_basamak += kimlikNoArr[i]; 
            }
            tum_basamaklar += kimlikNoArr[i]; // butun rakamlarida tum_basamaklar degiskeninde topluyoruz.
        }
        
        // kimlik numarasinin 10. hanesini matematiksel formul ile hesapliyoruz.
        kimlikNoArr[9] = (((tek_basamak) * 7) - (cift_basamak) ) % 10; 
        // hesaplanan 10.haneyide tum_basamaklar degiskenine ekliyoruz.
        tum_basamaklar += kimlikNoArr[9];
        // kimlik numarasinin son hanesini tum basamaklara uygulanan matematiksel islem ile buluyoruz.
        kimlikNoArr[10] = tum_basamaklar % 10;
        
        for(int i=0;i<11;i++){ // dizideki haneleri string olarak birlestiriyoruz.
            kimlikNo += kimlikNoArr[i];
        }
    }
    
    /*
     * disaridan KimlikNo cagrilmak istendiginde kullanilacak fonksiyon
     */
    public String getKimlikNo(){
        return kimlikNo;
    }
    
    /*
     * uretilmis kimlik numaralarini kontrol etmemizi saglayan fonksiyon
     * @param _kimlikNo kontrol edilecek kimlik numarasi
     */
    public static boolean kontrol(String _kimlikNo){
        // eger kimlik numarasinin ilk hanesi 0 ise diger islemleri yapmaya gerek olmadan
        // bu kimlik numarasinin yanlis oldugunu soyleyebiliriz bu yuzden false donduruyoruz.
        if(Character.getNumericValue(_kimlikNo.charAt(0)) == 0){
            return false;
        } 
                
        int cift_basamak = 0; // cift basamaktaki rakamlarin toplamini tutmak icin olusturulan degisken. 
        int tek_basamak = 0; // tek basamaktaki rakamlarin toplamini tutmak icin olusturulan degisken.
        int tum_basamaklar = 0; // tum basamaklardaki rakamlarin toplamini tutmak icin olusturulan degisken
        for(int i=0;i<9;i++){ // ilk 9 hanesi icin islem yapiyoruz.
            if(((i+1)%2) == 0){ // cift hanelerdeki rakamlari cift_basamak degiskeninde topluyoruz.
                cift_basamak += Character.getNumericValue(_kimlikNo.charAt(i)); 
            } else { // tek hanelerdeki rakamlari tek_basamak degiskeninde topluyoruz.
                tek_basamak += Character.getNumericValue(_kimlikNo.charAt(i)); 
            }
            tum_basamaklar += Character.getNumericValue(_kimlikNo.charAt(i)); // butun rakamlarida tum_basamaklar degiskeninde topluyoruz.
        }
                
        // kimlik numarasinin 10. hanesini matematiksel formul ile hesapliyoruz.
        int onBasamak = (((tek_basamak) * 7) - (cift_basamak) ) % 10; 
        // hesaplanan 10.haneyide tum_basamaklar degiskenine ekliyoruz.
        tum_basamaklar += onBasamak;
         // kimlik numarasinin son hanesini tum basamaklara uygulanan matematiksel islem ile buluyoruz.
        int sonBasamak = tum_basamaklar % 10;

        // eger kimlik numarasinin 10.hanesi bizim hesapladigimiz 10.haneye esit degilse kimlik numarasi yanlistir.
        if(Character.getNumericValue(_kimlikNo.charAt(9)) != onBasamak){
            return false;
        } else if(Character.getNumericValue(_kimlikNo.charAt(10)) != sonBasamak){
            // eger kimlik numarasinin son hanesi bizim hesapladigimiz son haneye esit degilse kimlik numarasi yanlistir.
            return false;
        }
        
        // eger program tum adimlari gecip buraya geldiyse o kimlik numarasi dogrudur bu yuzden true donduruyoruz
        return true;
    }
}
